src/Google.Cloud.Functions.Invoker/FunctionEnvironment.cs

```diff
@@ -149,8 +151,31 @@ internal static Type FindDefaultFunctionType(params Type[] types)
             };
 
             bool IsFunctionClass(Type t) =>
-                t.IsClass && !t.IsAbstract &&
-                (typeof(IHttpFunction).IsAssignableFrom(t) || typeof(ICloudEventFunction).IsAssignableFrom(t));
+                t.IsClass && !t.IsAbstract && !t.IsGenericType &&
+                (typeof(IHttpFunction).IsAssignableFrom(t) ||
+                typeof(ICloudEventFunction).IsAssignableFrom(t) ||
+                GetGenericInterfaceImplementationTypeArgument(t, typeof(ILegacyEventFunction<>)) is object);
+        }
+
+        /// <summary>
+        /// Checks whether a given type implements a generic interface (assumed to have a single type parameter),
+        /// and returns the type argument if so. For example, if we have "class MyFunction : ILegacyEventFunction{PubSubMessage}"
+        /// then a call to GetGenericInterfaceImplementationTypeArgument(typeof(MyFunction), typeof(ILegacyEventFunction&lt;&gt;))
+        /// will return typeof(PubSubMessage).
+        /// </summary>
+        /// <param name="target">The target type to check.</param>
+        /// <param name="genericInterface">The generic interface</param>
+        /// <returns></returns>
+        private static Type? GetGenericInterfaceImplementationTypeArgument(Type target, Type genericInterface)
```

Comment:
If `target` were to implement the `genericInterface` twice with different type arguments, we would rely on `GetInterfaces()` being deterministic in its order for always returning the same type argument, which I think we shouldn't. Maybe we can order by type argument name and always return the first one.
For instance:
`class MultiLegacyEventFunction : ILegacyEventFucntion<StorageObject>, ILegacyEventFunction<PubSubMessage> {...}`
We should always return the same one of `StorageObject` or `PubSubMessage`.
Very very ideally, `LegacyEventAdapter` could handle two (many) payload types, but I do think that might be a lot of work for the very few times it will happen.
But being deterministic in what we return is easy enough to implement and it's really nice to have.

Reply:
I think we should fail in that case. We wouldn't know which adapter to route it to, without extra information. Will implement and add a test.
(I don't think the error message needs to be super-clear here, as it's extremely unlikely to happen. I'm happy to field issues on that front.)